File: Swift/LeetCode/LeetCode396.swift
// https://leetcode.com/problems/rotate-function

class Solution {
    func maxRotateFunction(_ nums: [Int]) -> Int {
        let n = nums.count
        let totalSum = nums.reduce(0) { $0 + $1 }
        var current = nums.enumerated().reduce(0) { $0 + $1.offset * $1.element }
        var answer = current
        for i in 1..<n {
            current = current + totalSum - (n * nums[n - i]) 
            answer = max(answer, current)
        }
        return answer
    }
}

File: python/leetcode/leetcode396.py
# https://leetcode.com/problems/rotate-function

class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        n = len(nums)
        totalSum = sum(nums)
        current = sum(i * nums[i] for i in range(n))
        answer = current
        for i in range(1, n):
            current = current + totalSum - (n * nums[n - i])
            answer = answer if answer > current else current
        return answer
